### PR TITLE
Get the request.client under case uvicorn is run with a fd or a unix socket

### PR DESCRIPTION
**Pull request by euri10**
should solve #727 and still give the correct info that https://github.com/encode/uvicorn/issues/634 wanted

### CODE REVIEW

File: uvicorn/protocols/utils.py [hunk 1]
```diff
@@ -20,9 +15,17 @@ def get_remote_addr(transport):
...
-
+        elif hasattr(socket, "AF_UNIX") and family is socket.AF_UNIX:
+            if isinstance(info, tuple):
+                # fd case
+                # <uvloop.PseudoSocket fd=13, family=AddressFamily.AF_UNIX, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 8000), raddr=('127.0.0.1', 38634)>
+                return (str(info[0]), int(info[1]))
+            else:
+                # unix socket case
+                # <uvloop.PseudoSocket fd=21, family=AddressFamily.AF_UNIX, type=SocketKind.SOCK_STREAM, proto=0, laddr=/tmp/gunicorn.sock>
+                return None
```
Comment (florimondmanca) [inline uvicorn/protocols/utils.py]: I think I'd do something like this, to simplify:

```python
HAS_AF_UNIX = hasattr(socket, "AF_UNIX")

def get_remote_addr(transport):
    ...

        if family in (socket.AF_INET, socket.AF_INET6) or (
            HAS_AF_UNIX and family == socket.AF_UNIX and isinstance(family, tuple)
        ):
            return (str(info[0]), int(info[1]))

        return None
```

**Review by florimondmanca — COMMENTED** [review]
My understanding is that in some cases the `socket_info.getpeername()` may return a string containing the path to the Unix socket, and we need to ignore that case, correct?

I'm curious - in what cases, when bound to a Unix socket, can it return a (host, port) tuple? Isn't binding to a Unix socket not done via a host/port by definition? Had we confirmed that #636 _did fix_ #634?

**Comment by florimondmanca** [comment]
We might want to simply revert #636. It's not clear to me that it did fix anything, given how according to https://docs.python.org/3/library/socket.html there is no host/port information available when binding to a Unix socket. I think #634 was a non-issue ("yes there's no host/port because your server is not bound to a TCP socket").

**Comment by euri10** [comment]
> My understanding is that in some cases the `socket_info.getpeername()` may return a string containing the path to the Unix socket, and we need to ignore that case, correct?

correct

> I'm curious - in what cases, when bound to a Unix socket, can it return a (host, port) tuple? Isn't binding to a Unix socket not done via a host/port by definition? Had we confirmed that #636 _did fix_ #634?

if you run uvicorn with `supervisord -n` using

```
[supervisord]

[fcgi-program:uvicorn]
socket=tcp://localhost:8000
command=venv/bin/uvicorn --fd 0 app:app
process_name=uvicorn-%(process_num)d
stdout_logfile=/dev/stdout
stdout_logfile_maxbytes=0
```

then you end up having `socket_info` to what I left in comments ` <uvloop.PseudoSocket fd=13, family=AddressFamily.AF_UNIX, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 8000), raddr=('127.0.0.1', 38634)>`

**Comment by florimondmanca** [comment]
I see, so in that case Supervisord's [FastCGI support](http://supervisord.org/configuration.html#fcgi-program-x-section-settings) makes the TCP socket available at `fd=0`, which _is_ a Unix socket, _and_ for some reason the original TCP host:port info is made available on the Unix socket object too? (I'm definitely no Unix networking wizard, so trying to get my head around this…)

File: uvicorn/protocols/utils.py (hunk 1 above)
Comment (lovelydinosaur) [reply uvicorn/protocols/utils.py]: Could we just change the whole block to...

```python
return (str(info[0]), int(info[1])) if isinstance(info, tuple) else None
```

**Comment by bukowa** [comment]
I merged this patch into integration test with nginx - it works https://github.com/bukforks/uvicorn/tree/django_nginx
https://github.com/encode/uvicorn/pull/734

File: uvicorn/protocols/utils.py (hunk 1 above)
Comment (euri10) [reply uvicorn/protocols/utils.py]: I did it that way, way simpler,
Just note that I kept in `get_remote_addr` the try except introduced it seems in #495

**Review by florimondmanca — APPROVED** [review]

**Comment by florimondmanca** [comment]
I guess we can merge this and get it along for 0.11.8. 👍 #731

**Pull request merged**